service-coordtransform/src/main/java/fi/nls/oskari/spring/CoordinateTransformationAsyncController.java

```diff
@@ -0,0 +1,226 @@
+                handleException((Exception)res, params);
+                worker.clearJob(jobId);
+                return;
+            }
+            writeResponse(params, transformParams, (CoordinatesPayload)res);
+            worker.clearJob(jobId);
+        } catch (Exception e) {
+            handleActionException(e, params);
+        }
+    }
+
+    private void writeResponse(ActionParameters params, TransformParams transformParams,
+                               CoordinatesPayload coords) throws ActionException {
+        String targetCrs = transformParams.targetCRS;
+        int targetDimension = transformParams.outputDimensions;
+
+        HttpServletResponse response = params.getResponse();
+        if (transformParams.type.isFileOutput()) {
+            String fileName = addFileExt(transformParams.exportSettings.getFileName());
+            response.setContentType(FILE_TYPE);
+            response.setHeader("Content-Disposition", "attachment; filename=" + fileName);
+        } else {
+            response.setContentType(IOHelper.CONTENT_TYPE_JSON);
+        }
+        try (OutputStream out = response.getOutputStream()) {
+            if (transformParams.type.isFileOutput()) {
+                fileHelper.writeFileResponse(out, coords.getCoords(), targetDimension, coords.getExportSettings(), targetCrs);
+            } else {
+                writeJsonResponse(out, coords.getCoords(), targetDimension, coords.hasMore());
+            }
+        } catch (IOException e) {
+            throw new ActionException("Failed to write JSON to client", e);
+        }
+    }
+
+    private void handleException(Exception e, ActionParameters params) {
+        if (e instanceof ActionParamsException) {
+            handleActionParamsException((ActionParamsException)e, params);
+        } else {
+            handleActionException(e, params);
+        }
+    }
+
+    private void handleActionException(Exception e, ActionParameters params) {
+        Throwable error = e;
+        if(e.getCause() != null) {
+            error = e.getCause();
+        }
+        log.error(error,
+                "Couldn't handle action:", ROUTE,
+                "Message: ", e.getMessage(),
+                ". Parameters: ", params.getRequest().getParameterMap());
+        writeErrorResponse(params.getResponse(), e.getMessage(),  HttpServletResponse.SC_INTERNAL_SERVER_ERROR, null);
+    }
+
+    private void handleActionParamsException(ActionParamsException e, ActionParameters params) {
+        log.error("Couldn't handle action: " + ROUTE,
+                "Message: ", e.getMessage(),
+                ". Parameters: ", params.getRequest().getParameterMap());
+        writeErrorResponse(params.getResponse(), e.getMessage(),  HttpServletResponse.SC_NOT_IMPLEMENTED, e.getOptions());
+    }
+
+    //add .txt if missing
+    private String addFileExt(String name) {
+        int i = name.lastIndexOf('.');
+        if (i < 0 || i + 1 == name.length()) {
```

Comment:
If last char is dot we still add another dot before the file extension.

Reply:
Maybe we can simplify this whole method. Check if name ends with .$EXT and if it doesn't then add the .$EXT. But we probably still have to handle the special case when (while) the last char is a dot.